sky/spot/controller.py

```diff
@@ -309,17 +310,25 @@ def start(job_id, task_yaml, retry_until_up):
             time.sleep(1)
     except exceptions.SpotUserCancelledError:
         logger.info(f'Cancelling spot job {job_id}...')
+        cancelling = True
+    finally:
         if controller_process is not None:
-            logger.info('Killing controller process '
-                        f'{controller_process.pid}')
+            logger.info(f'Killing controller process {controller_process.pid}')
+            while controller_process.is_alive():
+                # NOTE: it is ok to kill or join a killed process.
+                controller_process.kill()
+                # Kill the children processes launched by log_lib.run_with_log.
+                subprocess_utils.kill_children_processes()
+                controller_process.join(timeout=1)
+
+        # TODO(suquark): It could take a long time cleaning up the cluster,
+        #  and in our current smoke tests it only waits for 5s after sending
+        #  the cancellation signal. So we set the spot cancellation status
+        #  before taking down the cluster (like what we did previously).
+        #  In the future, we may add a "cancelling" state for the spot
+        #  controller.
```

Comment:
Adding `CANCELLING` status is a good idea, but could we still have the `set_cancelled` after the `_clean_up` and set the wait time in the smoke tests longer?

Reply:
ok, let me fix it in this PR

Reply:
Oh, I think we can add the `CANCELLING` in a future PR if that needs significant changes, but we can just move the `set_cancelled` after the `_cleanup`. It is fine to have wait longer before the cluster is set to CANCELLED, because when a user call `sky spot cancel`, it will show the following:
```
I 03-15 05:44:38 spot_utils.py:164] Cancelling jobs 29.
Job with ID 29 is scheduled to be cancelled.
```
"scheduled to" indicates that the cancel will not happen immediately.

Reply:
got it, so I will:

1. move `set_cancelled` after _cleanup
2. extend the sleep time for the smoke tests before `grep CANCELLED` to ~200s (should be enough for taking down the clusters)
3. run the tests again to see if this works